Add tolerance to row count match

tylerriccio33 opened the pull request:
I added a tolerance to the row count match validation. It is useful for source data that is updating constantly or very large, ie. big time streaming data. I want the ability to allow for some tolerance.

rich-iannone commented on pointblank/validate.py:
```diff
@@ -4551,6 +4551,7 @@ def col_schema_match(
     def row_count_match(
         self,
         count: int | FrameT | Any,
+        tol: float = 0,
```
If we're adding a `tol=` arg (which is totally reasonable) could you add a description of it in the docstring? Also, it might be good to have the option to input as a tuple/list for setting lower/upper bounds.

tylerriccio33 replied:
I totally missed the updating the docstring - my mistake! As for lower/upper bounds, I can use the suggestion you had: if 0-1 represent tolerance as a percentage of the actual count, if tuple then treat as raw limits and if int above 1 treat as literal limit.

rich-iannone commented on tests/test_validate.py:
```diff
@@ -2660,6 +2661,25 @@ def test_row_count_match(request, tbl_fixture):
+                         ])
+def test_row_count_tol(nrows : int, target_count: int, tol : float, should_pass : bool) -> None:
+
+    rng = np.random.default_rng()
```
Loving the tests, but would it be possible to get away with not using numpy?

tylerriccio33 replied:
Sure thing!

rich-iannone reviewed (changes requested):
This is a great idea, but I think we should work a bit more on getting the implementation down. Left a few comments, we can discuss a bit more in the Conversation view.

rich-iannone commented:
One more idea that could be interesting would be to accept/process values similar to the `Threshold` class: any value between between `0`-`1` is relative, `1` and above values would be absolute rows.

tylerriccio33 commented:
New pr has the below:
- removed numpy as the rng and went back to good ole' random. I'm opening a separate pr on using hypothesis for parametric testing instead, details there.
- added ability to pass literal limits as a tuple, a literal limit as int and a relative limit as in the original pr. This takes a bit of validation.
- Added new examples to demonstrate plus a test of those examples.

rich-iannone commented:
Thanks! Will review this soon :)

rich-iannone commented:
I feel like we're *very* close here to enhancing this method. One last thing: I think the rules for relative/absolute limits should apply to both the scalar case *and* the 2-item tuple case. 

The basic case (scalar) would be used most often and just means that the range is equal on both sides (so using `tol=0.1` could expand to, for the internal logic, `tol=(0.1, 0.1)`). And I can see using a mix of absolute and relative bounds for some cases (e.g., there should be at least x rows, but perhaps up to 10% were added; so: `tol=(0, 0.1)`).

If this is okay, I'd recommend (in the code) upgrading any scalar value to a tuple (`5` -> `(5, 5`)) and always handling the lower/upper bounds. Sorry to keep adding work, but this will be so useful once completed!

rich-iannone commented on pointblank/_interrogation.py:
```diff
@@ -1893,16 +1893,31 @@ class RowCountMatch:
 
-        if not self.inverse:
-            res = get_row_count(data=self.data_tbl) == self.count
+        row_count :int = get_row_count(data=self.data_tbl)
```
Is it common/better to declare types for variables? I'm honestly not sure what the norm is on this one. I only ask because if it is a preferred approach, we should probably be doing this everywhere (which is what I'm leaning toward).

tylerriccio33 replied:
"best practice" says to type hint everything but it's not necessary. Implementing type checking on a codebase of this size is a giant task, worth it in the long run but not trivial. I'd be happy to do so if you're on board though.

I will say because of how dynamic this code base is, it would probably be even hard than usual, but not at all impossible.

The most bang-for-buck would be to type check function signatures, it's 10% of the work for 90% gain. In my experience, so take it with a grain of salt, most things are caught at the signature level and not the individual variable level.

rich-iannone replied:
Thanks for weighing in on this. I think we ought to add types to variable declarations! We can both chip away at this, and I think it'll be worth it.

tylerriccio33 commented on Makefile:
```diff
@@ -7,7 +7,7 @@ test-update:
 
 test-coverage:
-	pytest --cov=pointblank --cov-report=term
+	pytest --cov=pointblank --cov-report=term-missing
```
I accidentally left that in the pr 😂

tylerriccio33 commented:
> I feel like we're _very_ close here to enhancing this method. One last thing: I think the rules for relative/absolute limits should apply to both the scalar case _and_ the 2-item tuple case.
> 
> The basic case (scalar) would be used most often and just means that the range is equal on both sides (so using `tol=0.1` could expand to, for the internal logic, `tol=(0.1, 0.1)`). And I can see using a mix of absolute and relative bounds for some cases (e.g., there should be at least x rows, but perhaps up to 10% were added; so: `tol=(0, 0.1)`).
> 
> If this is okay, I'd recommend (in the code) upgrading any scalar value to a tuple (`5` -> `(5, 5`)) and always handling the lower/upper bounds. Sorry to keep adding work, but this will be so useful once completed!

tylerriccio33 commented:
So I dedicated a type to the bounds which should solve the validation piece for people who use type checkers. I abstracted the validation and derivation of the left and right bounds into a separate function. I think this should cover all cases and the approach will allow extension in the future and multiple uses of the new type.

rich-iannone commented on pyproject.toml:
```diff
@@ -122,3 +122,8 @@ testpaths = [
 
 [tool.black]
 line-length = 100
+
```
I added this coverage exclusion. Hope it's alright!

rich-iannone reviewed (approved):
LGTM!
